### PR TITLE
FISH-341 Create GCP Secrets Cloud Config Source

### PR DESCRIPTION
**Pull request by MattGill98**
## Description

This is a new feature, which looks to allow GCP secrets manager to be used as a config source for Payara.

## Documentation

https://github.com/payara/Payara-Community-Documentation/pull/88

## Notes for Reviewers

Note that although this doesn't perfectly follow the design for an extension, the design is such that moving this provider to an extension would be quite simple. As such a new service is implemented that registers and dynamically updates the config sources. A proxy class is injected to ensure that disabling a config source disables it without requiring a restart

Currently only the JSON key file (and not the p12) is supported, and it's copied into the config directory for synchronisation. Note that in future, it would be good to encrypt these files in a similar fashion to the `datagrid-key` file.

The OAuth2 client in the oauth security module wasn't quite what was needed, so I've implemented a new OAuth2 client and made both modules use it. The OAuth2 security module might need testing to confirm it still works as intended.

### CODE REVIEW

**Comment by MattGill98** [comment]
jenkins test please

**Comment by MattGill98** [comment]
jenkins test please

**Comment by MattGill98** [comment]
jenkins test please

**Comment by MattGill98** [comment]
jenkins test please

File: appserver/payara-appserver-modules/microprofile/config-extensions/pom.xml [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+  only if the new code is made subject to such option by the copyright
+  holder.
+-->
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+
+    <parent>
+        <groupId>fish.payara.server.internal.payara-appserver-modules</groupId>
+        <artifactId>microprofile</artifactId>
+        <version>5.2020.5-SNAPSHOT</version>
```
Comment (MattGill98) [inline appserver/payara-appserver-modules/microprofile/config-extensions/pom.xml:47]:
```suggestion
        <version>5.2020.6-SNAPSHOT</version>
```

**Review by MattGill98 — COMMENTED** [review]
Needs updating to Snapshot

**Comment by MattGill98** [comment]
jenkins test please

File: appserver/payara-appserver-modules/microprofile/config-extensions/src/main/java/fish/payara/microprofile/config/extensions/gcp/GetGCPSecretsConfigSourceConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+import org.glassfish.api.admin.RuntimeType;
+import org.glassfish.config.support.CommandTarget;
+import org.glassfish.config.support.TargetType;
+import org.glassfish.hk2.api.PerLookup;
+import org.jvnet.hk2.annotations.Service;
+
+import fish.payara.nucleus.microprofile.config.source.extension.BaseGetConfigSourceConfigurationCommand;
+import fish.payara.nucleus.microprofile.config.spi.MicroprofileConfigConfiguration;
+
+@Service(name = "get-gcp-secrets-config-source-configuration")
```
Comment (MeroRai) [inline appserver/payara-appserver-modules/microprofile/config-extensions/src/main/java/fish/payara/microprofile/config/extensions/gcp/GetGCPSecretsConfigSourceConfiguration.java:57]: The command is too long in my opinion. I suggest shorting it, for examples: `get-gcp-secrets-configuration`, `get-gcp-config` or `get-gcp-configuration`

File: appserver/payara-appserver-modules/microprofile/config-extensions/src/main/java/fish/payara/microprofile/config/extensions/gcp/SetGCPSecretsConfigSourceConfigurationCommand.java [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+import org.glassfish.api.admin.ServerEnvironment;
+import org.glassfish.config.support.CommandTarget;
+import org.glassfish.config.support.TargetType;
+import org.glassfish.hk2.api.PerLookup;
+import org.jvnet.hk2.annotations.Service;
+
+import fish.payara.nucleus.microprofile.config.source.extension.BaseSetConfigSourceConfigurationCommand;
+import fish.payara.nucleus.microprofile.config.spi.MicroprofileConfigConfiguration;
+
+@Service(name = "set-gcp-secrets-config-source-configuration")
```
Comment (MeroRai) [inline appserver/payara-appserver-modules/microprofile/config-extensions/src/main/java/fish/payara/microprofile/config/extensions/gcp/SetGCPSecretsConfigSourceConfigurationCommand.java:71]: Same as my above comment. I suggest shorting the command name

**Review by MeroRai — CHANGES_REQUESTED** [review]
Looks good. I left suggestions which you can take into consideration. I noticed some of the comments and code still refer to the Notifier service this will need updating. Finally, some of the class doesn't have a copyright header.

**Comment by MattGill98** [comment]
jenkins test please

**Comment by MattGill98** [comment]
jenkins test please

**Review by MeroRai — APPROVED** [review]

**Comment by MattGill98** [comment]
jenkins test please

**Pull request merged**